### PR TITLE
refactor(ci): Remove reflection from Cis class

### PR DESCRIPTION
**Pull request by ezimanyi**
* refactor(ci): Move CI service and controller their own package 

* refactor(ci): Combine CI+Master for Controller and Service 

  In the next commit we'll be subclassing these for each type of CI; it makes more sense to split the files along the lines of CI type than to have separate classes for the CI and the Master.

* refactor(ci): Make base Controller + Service abstract 

  In order to support static typing of CI classes, make the base controller and service abstract, and have each type of CI integration extend that class and supply the particular type of CI.

* refactor(ci): Remove reflection from Cis class 

  Now that the services and controllers use static typing, we can remove the reflection to dynamically determine the class to build.

* refactor(ci): Rename base class from Master -> CIAccount 

  The term 'Master' is very Jenkins-specific and does not apply to more general build systems. We'll leave the existing subclasses as is, but will rename the base class to CIAccount to better reflect that this represents a general account rather than a particular server.

### CODE REVIEW

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/ci/CiService.java [hunk 1]
```diff
@@ -18,53 +18,53 @@
...
 
 /**
  * This service is meant to be autowired into any service or controller that needs to inspect the current halconfigs
  * cis.
  */
-@Component
-public class CiService {
-  @Autowired
-  private LookupService lookupService;
+@RequiredArgsConstructor
```
Comment (jtk54) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/ci/CiService.java]: If all the sub-classes are meant to be spring beans, I think you can leave the `@Component` here.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/ci/ConcourseService.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+@Component
+public class ConcourseService extends CiService<ConcourseMaster, ConcourseCi> {
+    public ConcourseService(LookupService lookupService, ValidateService validateService) {
+        super(lookupService, validateService);
+    }
+
+    public String ciName() {
+        return "concourse";
+    }
+
```
Comment (jtk54) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/ci/ConcourseService.java]: Extra line.

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/ci/CiController.java [hunk 3]
```diff
@@ -31,129 +30,118 @@
...
+import org.springframework.stereotype.Component;
 import org.springframework.web.bind.annotation.*;
 
 import java.util.List;
 
-@RestController
-@RequiredArgsConstructor
-@RequestMapping("/v1/config/deployments/{deploymentName:.+}/ci")
-public class CiController {
+public abstract class CiController<T extends Master, U extends Ci<T>> {
```
Comment (jtk54) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/ci/CiController.java]: Little confused about what's going on here with the API root (e.g. `/v1/config/deployments/...`). Can you clarify?

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/ci/ConcourseController.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+package com.netflix.spinnaker.halyard.controllers.v1.ci;
+
+import com.netflix.spinnaker.halyard.config.model.v1.ci.concourse.ConcourseCi;
+import com.netflix.spinnaker.halyard.config.model.v1.ci.concourse.ConcourseMaster;
+import com.netflix.spinnaker.halyard.config.services.v1.ci.ConcourseService;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/v1/config/deployments/{deploymentName:.+}/ci/concourse")
```
Comment (jtk54) [inline halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/ci/ConcourseController.java]: OK nevermind --  I get it now.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/ci/CiService.java (hunk 1 above)
Comment (ezimanyi) [reply halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/ci/CiService.java]: I just tried this now, and it didn't seem like Spring was creating beans for the subclasses.

Comment (jtk54) [reply halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/ci/CiService.java]: I stand corrected, must have misremembered that.

File: halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/ci/CiController.java (hunk 3 above)
Comment (ezimanyi) [reply halyard-web/src/main/java/com/netflix/spinnaker/halyard/controllers/v1/ci/CiController.java]: Each sub-controller is declaring a specific path for that CI system.

**Pull request merged**